fs: remove permissive rmdir recursive

aduh95 opened the pull request:
To follow the plan invoked in https://github.com/nodejs/node/pull/35562#issuecomment-706419155.

This removes permissive behaviour of `fs.rmdir(path, { recursive: true })`, and add a runtime warning for all use of `recursive` option with `rmdir` (currently, the warning is emitted only if `path` isn't a directory).

<!--
Before submitting a pull request, please read
https://github.com/nodejs/node/blob/master/CONTRIBUTING.md.

Commit message formatting guidelines:
https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

iansu commented:
Hey, thanks for doing this. A couple of initial thoughts:

1. In terms of the new behaviour of `rmdir` we should probably just put it back to what it was before the `recursive` option was introduced. I haven't had a chance to go back and look at that code yet but that might help with your question about what error to throw.
2. We want this change to go into Node 16 but I'm not sure how to make that happen. I think the `semver-major` label you've applied will do that but I'm not 100% sure.

bcoe commented:
@aduh95, just had a conversation with @iansu in slack. I dug up this summary of the conversation we had with the TSC:

https://github.com/nodejs/node/pull/35494#issuecomment-703290650

My temptation is that we continue supporting `rmdir(path, {recursive: true})`, but that it throws on a missing or file path (this brings us in alignment with APIs in Deno, Python, and I believe C#).

If we then later decide it's silly to maintain both implementations, we can later deprecate the `recursive` option.

I think there's a danger of removing `recursive` completely breaking the world for people in Node 16, whereas moving towards being stricter will break far fewer people, and bring us closer to the APIs on other platforms.

aduh95 commented:
@bcoe That was my understanding as well, and I believe this PR is doing what you described above: the `recursive` option is still there, but throws if the `path` is not a directory.

iansu commented:
@aduh95 Sorry for sending you off in the wrong direction. I do think it's worth considering if we want to also print a deprecation warning for *any* use of rmdir recursive in Node 16. That might be a bit more controversial though so I think that's best discussed in a followup issue or PR.

bcoe commented:
> I do think it's worth considering if we want to also print a deprecation warning for any use of rmdir recursive in Node 16

@iansu this sounds reasonable to me, it might honestly be worth bringing up to the TSC, i.e., do we want to permanently support the `rm` helper, and a recursive option on `rmdir`, or do we want to nudge people towards `rm`.

aduh95 commented:
> it might honestly be worth bringing up to the TSC

I agree! This PR is marked `semver-major`, that means it needs at least two approvals from TSC members. Before pinging them, I'd like to make sure the work I've done aligns with plan we can agree on. It seems to me there are two things conflicting and I'm not sure I've got it right:

1. Deprecation status of `recursive` option: do we want to remove the option for Node.js 17 in favor of `fs.rm`? I would say that yes, we should align with `RMDIR(1)`, which doesn't have a `recursive` option.
2. Permissive behaviour of `fs.rmdir(path, {recursive:true})`: we want to get rid of it, but is it worth it if we are planning to remove the `recursive` option anyway?

Do you think it's a good summary?

> My temptation is that we continue supporting `rmdir(path, {recursive: true})`, but that it throws on a missing or file path (this brings us in alignment with APIs in Deno, Python, and I believe C#).

If we want to align with the other APIs, shouldn't we want to remove the option instead?

- Deno doesn't have a `rmdir` API AFAICT, [`Deno.remove`](https://doc.deno.land/builtin/stable#Deno.remove) has a `recursive` option but is more analogous to `fs.rm`.
- Python doesn't have a `recursive` option it seems: https://docs.python.org/3/library/os.html#os.rmdir.
- C# `RmDir` doesn't have a `recursive` option either: https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualbasic.filesystem.rmdir?view=net-5.0

If that's OK with you, I'm going to open another PR that adds a runtime warning for all uses of `recursive` option with `rmdir[Sync]` to nudge people to move to `fs.rm` instead, as it seems the "correct" way of action.

bcoe commented on test/parallel/test-fs-rmdir-recursive-throws-not-found.js:
```diff
@@ -0,0 +1,36 @@
+    })
+  );
+}
+{
```
mind adding a comment above each block describing the test case?

aduh95 replied:
There all the same test case, one for each `rmdir` method. Do you have suggestions for comments that are not redundant?

```js
// Testing fs.rmDirSync.
// Testing fs.rmDir.
// Testing fs.promises.rmDir.
```

bcoe replied:
I suppose you're right, the comments can be omitted.

bcoe reviewed (approved):
👍 with one nit, mind adding a comment above the blocks of your test cases, describing each?

aduh95 commented:
I've rebased on top of master to solve the conflict, and also changed deprecation status from `End-of-Life` to `Documentation only`: indeed the permissive behavior is EOL, but my understanding is that the `recursive` option is still deprecated, and we want to remove it in a future version of Node.js. /cc @bcoe @jasnell in case you'd want to retract your approval.

bcoe commented on doc/api/deprecations.md:
```diff
@@ -2668,19 +2668,25 @@ The [`crypto.Certificate()` constructor][] is deprecated. Use
+and `fs.promises.rmdir(path, { recursive:true })` throws
 if `path` does not exist or is a file.
-Use `fs.rm(path, { recursive: true, force: true })` instead.
+Use `fs.rm(path, { recursive: true, force: true })`,
```
This still mainly refers to the permissive behavior, which is removed in `v16.0.0`.

Wondering if it would get the point across better if we went with:

```md
`fs.rmdir(path, { recursive: true })`, `fs.rmdirSync(path, { recursive:true })` will be removed in the next major, use Use `fs.rm(path, { recursive: true, force: true })`, etc.
```

I'm 👍 with this decision as long as the TSC is supportive -- avoid keeping both API surfaces.

Trott replied:
I don't think we want to say it will be removed in 17.x. "a future version" is sufficient. "the next version" is more specific than we've been in the past, and we've been known to kick things to subsequent majors for various reasons.

Trott replied:
Oh, wait, it says "a future version" now. Cool.

bcoe reviewed (changes requested):
Left a nit regarding the wording of the deprecation notice, given that we're leaning towards removing recursive entirely in `v17.0.0`.

bcoe commented:
@mhdawson would it be possible to put this on a @nodejs/tsc agenda for visibility, I feel @aduh95 has proposed a good path forward:

1. remove the permissive behavior from `rs.rmdir` in `v16` (as discussed).
2. documentation only deprecation of `rs.rmdir(path, {recursive: true})` in `v16`.
3. In `v17` recursive behavior is dropped from `rmdir` entirely, in favor of `fs.rm`.

iansu commented on doc/api/fs.md:
```diff
@@ -4778,8 +4802,8 @@ changes:
-    recursive mode, errors are not reported if `path` does not exist, and
-    operations are retried on failure. **Default:** `false`.
+    recursive mode, operations are retried on failure. **Default:** `false`.
+    **Deprecated**.
```
It's a bit hard to read this diff. Does this say the `recursive` option is deprecated? That shouldn't be the case for `rm` but I might be reading this wrong.

aduh95 replied:
No, it says that `recursive` option of `fs.rmDirSync` is deprecated. `fs.rm` is actually 1000 lines above if you check the line number, but I agree the diff is kinda hard to read here.

aduh95 commented:
> 1. remove the permissive behavior from `rs.rmdir` in `v16` (as discussed).
> 1. documentation only deprecation of `rs.rmdir(path, {recursive: true})` in `v16`.
> 1. In `v17` recursive behavior is dropped from `rmdir` entirely, in favor of `fs.rm`.

I think we would need to go through a runtime deprecation first, so either:

- runtime deprecate + remove permissive behavior:
 1. remove the permissive behavior from `rs.rmdir` in `v16` (as discussed).
 1. runtime deprecation of `rs.rmdir(path, {recursive: true})` in `v16`.
 1. In `v17` recursive behavior is dropped from `rmdir` entirely, in favor of `fs.rm`.
- remove permissive behavior, deprecate later (that's the path taken by this PR):
 1. remove the permissive behavior from `rs.rmdir` in `v16` (as discussed).
 1. doc-only deprecation of `rs.rmdir(path, {recursive: true})` in `v16`.
 1. runtime deprecation of `rs.rmdir(path, {recursive: true})` in `v17`.
 1. In `v18` recursive behavior is dropped from `rmdir` entirely, in favor of `fs.rm`.
- runtime deprecate while keeping permissive behavior (that's the path taken by https://github.com/nodejs/node/pull/37302):
 1. DO NOT remove the permissive behavior from `rs.rmdir`, instead runtime deprecate all use of `rs.rmdir(path, {recursive: true})` in `v16`.
 1. In `v17` recursive behavior is dropped from `rmdir` entirely, in favor of `fs.rm`.

---

I feel the latter is the "less breaking path", as it avoids introducing a breaking change, but on the other hand the permissive behavior may be quite surprising to new users. No strong opinions though, I think all three paths are reasonable.

iansu commented:
@aduh95 I think I'm leaning towards option 1. If we are going to remove the recursive functionality from `rmdir` I'd prefer to do it as quickly as possible. I think the longer it stays the harder it will be to remove it fully.

mhdawson commented:
I see @aduh95 added the tsc-agenda tag, that will make sure it shows up on the agenda for the next meeting.

Trott commented:
CITGM (queued): https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/2647/

mhdawson reviewed (approved):
LGTM as following previously agreed plan

aduh95 commented:
Landed in 9948036ee050
